package.json

```diff
@@ -91,6 +93,8 @@
     "istanbul-instrumenter-loader": "^3.0.1",
     "jasmine-core": "^3.5.0",
     "jasmine-spec-reporter": "^7.0.0",
+    "jest": "^27.5.1",
```

Comment:
I'm not sure if this is what you were asking but I think it's from this Rust wasm-pack (https://github.com/rustwasm/wasm-pack) and the build.md in wasm-pack produces the binary artifact index_bg.wasm along with index.d.ts, index.js, and index_bg.d.ts (https://github.com/rustwasm/wasm-pack/blob/0227505a08cf86d7d27fa333510425cede4a3b30/docs/src/commands/build.md?plain=1#L48)

Reply:
Great, thanks for finding that! Can you just confirm that the `2e/native-esm/wasm-bindgen/index_bg.wasm` file in this dependency can be produced by `wasm-pack`? I looked in `wasm-pack` but couldn't find an example `index_bg.wasm` file to compare, so I think you might need to get `wasm-pack` running locally to do this test.

Some background: The concern here is that binary files can't be easily inspected by people reviewing the repository, so they're a great place to hide malicious code. For example, this technique was used to plant a backdoor in [xz utils](https://en.wikipedia.org/wiki/XZ_Utils_backdoor) that was discovered earlier this year. To determine that a binary file is safe, we need to either show that it's a known non-code file (e.g. an image so that malicious code trying to execute code inside the image would look very suspicious to reviewers) or that it can be generated from reviewable source code. In this case we have a web-assembly file, which is designed to hold code, so we're left with the second option. I determined `e2e/native-esm/add.wasm` was safe because it was a bit-for-bit match to the output of source code in https://github.com/mdn/webassembly-examples/blob/2f2163287f86fe29deb162335bccca7d5d95ca4f/understanding-text-format/add.wasm. We need to do something similar for `index_bg.wasm`.

Reply:
@U8NWXD 
I've been trying to figure it out but since I'm not familiar with these binary files, I'm still not sure if I'm checking it correctly. I tried making a new Rust project called `index` and `wasm-pack build --out-name index` to create the binary file that's also called index_bg.wasm. Then, I compare this `index_bg.wasm` and `e2e/native-esm/wasm-bindgen/index_bg.wasm` from jest and this is the result:

```
imchristie@Christie:~/jestest/index$ sha256sum pkg/index_bg.wasm
5c690301f57fc1b69fa1f968a6acbdc65383cbc86079e63b955c33363abce591  pkg/index_bg.wasm
imchristie@Christie:~/jestest/index$ sha256sum ~/jestest/jest/e2e/native-esm/wasm-bindgen/index_bg.wasm
d0cd69a345d190d63613acbb103b127f38ad2c3a435f5e37c2e22aaeecf68dea  /home/imchristie/jestest/jest/e2e/native-esm/wasm-bindgen/index_bg.wasm
```

When a new Rust project is created and "`wasm-pack build`" is run, a binary file is always created. I hope this is what you are asking for.

Reply:
@imchristie What happens if you try with a project that uses the files here, rather than a brand-new Rust project? https://github.com/rustwasm/wasm-bindgen/tree/4f865308afbe8d2463968457711ad356bae63b71/examples/hello_world -- does that make the sha256sum for both of them match?

The [source code](https://github.com/jestjs/jest/pull/13608/commits/942188c53e4caf1b0318c7c860e43934d2bc61cd#diff-3b83478538be228907c7873ff0db753bb7313a417dfd25db542b39e0a509bbe7R1) seems to indicate that the above example project is the origin for the code in the jest repo.

Reply:
Just an update -- I looked into this a bit. Here's what I found:

- The Jest code likely arises from running wasm-pack on https://github.com/rustwasm/wasm-bindgen/tree/4f865308afbe8d2463968457711ad356bae63b71/examples/hello_world.
- I ran `wasm-pack build` on the latest version of https://github.com/rustwasm/wasm-bindgen/tree/main/examples/hello_world and got some similar code for .js and _bg.js, but it wasn't exactly the same. I couldn't compare the .wasm files.
- The wasm-bindgen example code has undergone some changes since the date of the Jest PR. Newer versions of Rust don't work with the older wasm-bindgen crate from that time -- `wasm-pack build` fails with these errors:
    ```
    error: older versions of the `wasm-bindgen` crate will be incompatible with future versions of Rust; please update to `wasm-bindgen` v0.2.88
      |
      = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
      = note: for more information, see issue #71871 <https://github.com/rust-lang/rust/issues/71871>
    ```

Here is a copy of the hello_world.js and hello_world_bg.js files generated using this process. 

<details>
<summary>hello_world.js:</summary>
```
import * as wasm from "./hello_world_bg.wasm";
export * from "./hello_world_bg.js";
import { __wbg_set_wasm } from "./hello_world_bg.js";
__wbg_set_wasm(wasm);
wasm.__wbindgen_start();
```
</details>

<details>
<summary>hello_world_bg.js:</summary>

```
let wasm;
export function __wbg_set_wasm(val) {
    wasm = val;
}


const lTextDecoder = typeof TextDecoder === 'undefined' ? (0, module.require)('util').TextDecoder : TextDecoder;

let cachedTextDecoder = new lTextDecoder('utf-8', { ignoreBOM: true, fatal: true });

cachedTextDecoder.decode();

let cachedUint8ArrayMemory0 = null;

function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

let WASM_VECTOR_LEN = 0;

const lTextEncoder = typeof TextEncoder === 'undefined' ? (0, module.require)('util').TextEncoder : TextEncoder;

let cachedTextEncoder = new lTextEncoder('utf-8');

const encodeString = (typeof cachedTextEncoder.encodeInto === 'function'
    ? function (arg, view) {
    return cachedTextEncoder.encodeInto(arg, view);
}
    : function (arg, view) {
    const buf = cachedTextEncoder.encode(arg);
    view.set(buf);
    return {
        read: arg.length,
        written: buf.length
    };
});

function passStringToWasm0(arg, malloc, realloc) {

    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }

    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = encodeString(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}
/**
 * @param {string} name
 */
export function greet(name) {
    const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len0 = WASM_VECTOR_LEN;
    wasm.greet(ptr0, len0);
}

export function __wbg_alert_196a1d28b75a1fd4(arg0, arg1) {
    alert(getStringFromWasm0(arg0, arg1));
};

export function __wbindgen_init_externref_table() {
    const table = wasm.__wbindgen_export_0;
    const offset = table.grow(4);
    table.set(0, undefined);
    table.set(offset + 0, undefined);
    table.set(offset + 1, null);
    table.set(offset + 2, true);
    table.set(offset + 3, false);
    ;
};
```
</details>

The WASM file gives a sha256sum of 282735a8e3fd6fca7789e52c11c7ae9afaa5152a3969545ff9f8903d51b1101e which doesn't match anything in @imchristie's notes. 

Note that the _bg.js file here is a bit different from the [Jest one](https://github.com/jestjs/jest/pull/13608/files#diff-4091801a000c9f5ad62d0af63720625e02961282c67ec17c9674f9c0a9500cc0), though, and those Jest files haven't been updated since PR #13608 in Jan 2023 (except to add a copyright notice) -- see [link](https://github.com/jestjs/jest/tree/main/e2e/native-esm/wasm-bindgen). My guess is that the toolchain has changed since 2023 and we would need to find the correct versions of wasm-pack, cargo, rust, etc. from that time and run those.

That said, it's also worth noting that the jest/e2e directory seems to be just for integration test files: https://github.com/jestjs/jest/blob/main/CONTRIBUTING.md#integration-tests, so files within there are unlikely to affect users of the library.

@U8NWXD what are your thoughts on how to proceed, given the above? @imchristie if you have any suggestions or additional findings, please feel free to chime in too!

Reply:
I think incorporating this dependency is okay since the binary file clearly should not be referenced by production code. Specifically, it's in an e2e test directory, so production code referencing it would be suspicious. Here's my reasoning:

With any third-party dependency, we're taking on a risk that the third-party code might behave maliciously. This risk is mitigated somewhat by the fact that our dependencies are open-source and broadly used, so hopefully other developers would have noticed malicious code. This mitigation is weakened by binary files since these files cannot be easily reviewed by developers. In other words, if malicious code were hidden in a binary file, reviewers might not notice it.

However, malicious code in a binary file is only a problem if it gets executed somehow. In a project like jest where the entrypoints we execute have source code we can review, code in a binary can only be executed if some code that's not in a binary, and therefore is human-reviewable, executes it. This means that the risk of a binary file can be dramatically reduced if any code referencing that binary file would be obviously dangerous. For example, image files are binary, but that's okay because a reviewer should notice code trying to execute or load code from an image. Similarly, binary files in testing directories are okay because production code reading a test asset is obviously problematic.

I've updated the dependency incorporation checklist in https://github.com/oppia/oppia-web-developer-docs/issues/327 accordingly. Sorry for the delay @imchristie, but thanks for helping fine-tune our guidelines for adding dependencies!